From: Ivan Kvashonkin <vizzlef@gmail.com>
Date: Sun, 26 Jul 2026 17:00:18 +0300
Subject: [PATCH] client: set TCP_NODELAY on accepted sockets

Avoid delayed-ACK stalls between the multiple writes making up responses
on reused keep-alive connections. This is upstream commit
82b4c79ec266942251f95010eae700ddc33899ca.

Signed-off-by: Ivan Kvashonkin <vizzlef@gmail.com>
--- a/client.c
+++ b/client.c
@@ -20,6 +20,7 @@
 #include <libubox/blobmsg.h>
 #include <errno.h>
 #include <limits.h>
+#include <netinet/tcp.h>
 
 #include "uhttpd.h"
 #include "tls.h"
@@ -849,6 +850,7 @@ bool uh_accept_client(int fd, bool tls)
 	static struct client *next_client;
 	struct client *cl;
 	unsigned int sl;
+	int yes = 1;
 	int sfd;
 	static int client_id = 0;
 	struct sockaddr_in6 addr;
@@ -863,6 +865,10 @@ bool uh_accept_client(int fd, bool tls)
 	if (sfd < 0)
 		return false;
 
+	/* Avoid Nagle/delayed-ACK stalls between the separate writes used for
+	 * response headers and bodies on reused keep-alive connections. */
+	setsockopt(sfd, IPPROTO_TCP, TCP_NODELAY, &yes, sizeof(yes));
+
 	set_addr(&cl->peer_addr, &addr);
 	sl = sizeof(addr);
 	getsockname(sfd, (struct sockaddr *) &addr, &sl);
